### ISSUE TITLE
Invalid IL Code for Reconciliation

### ISSUE THREAD

**Issue by Lutz89** [FirstGearGames/FishNet#49]
FishNet 1.5.2 Pro
Unity 2021.3.2f1
Client and Server (Editor and Standalone Build)

Hey @FirstGearGames , I found an error that creates an invalid IL code for the Reconciliation method and thus causes the client / editor to crash. 
The crash happens when the client (not as host) connects.

Steps to reproduce:
1) Attach / Import script
2) Let code gen run
3) Check generated (IL) Code for Reconciliation

Decompiled Code in dotPeek:
```cs
[Reconcile]
private void Reconciliation(RigidbodyPrediction.ReconcileData rd, bool asServer)
{
      // ISSUE: unable to decompile the method.
}
```

Sample Code:
```cs
using FishNet;
using FishNet.Object;
using FishNet.Object.Prediction;
using UnityEngine;

namespace FishNet.Example.Prediction.Rigidbodies
{
    public interface TestObject
    {
        Vector3 Position { get; set; }
    }

    public class RigidbodyPrediction : NetworkBehaviour
    {
        #region Types.
        public struct MoveData
        {
            public bool Jump;
            public float Horizontal;
            public float Vertical;
            public MoveData(bool jump, float horizontal, float vertical)
            {
                Jump = jump;
                Horizontal = horizontal;
                Vertical = vertical;
            }
        }
        public struct ReconcileData
        {
            public Vector3 Position;
            public Quaternion Rotation;
            public Vector3 Velocity;
            public Vector3 AngularVelocity;
            public Vector3 ObjectPos;

            public ReconcileData(Vector3 position, Quaternion rotation, Vector3 velocity, Vector3 angularVelocity, Vector3 objectPos)
            {
                Position = position;
                Rotation = rotation;
                Velocity = velocity;
                AngularVelocity = angularVelocity;
                ObjectPos = objectPos;
            }
        }
        #endregion

        #region Serialized.
        [SerializeField]
        private float _jumpForce = 15f;
        [SerializeField]
        private float _moveRate = 15f;
        #endregion

        #region Private.
        private Rigidbody _rigidbody;
        private float _nextJumpTime;
        private bool _jump;

        // Sample object: e.g. a pickup object etc
        private TestObject _testObject = null; 
        #endregion

        private void Awake()
        {

            _rigidbody = GetComponent<Rigidbody>();
            InstanceFinder.TimeManager.OnTick += TimeManager_OnTick;
            InstanceFinder.TimeManager.OnPostTick += TimeManager_OnPostTick;
        }

        private void OnDestroy()
        {
            if (InstanceFinder.TimeManager != null)
            {
                InstanceFinder.TimeManager.OnTick -= TimeManager_OnTick;
                InstanceFinder.TimeManager.OnPostTick -= TimeManager_OnPostTick;
            }
        }

        private void Update()
        {
            if (base.IsOwner)
            {
                if (Input.GetKeyDown(KeyCode.Space) && Time.time > _nextJumpTime)
                {
                    _nextJumpTime = Time.time + 1f;
                    _jump = true;
                }
            }
        }

        private void TimeManager_OnTick()
        {
            if (base.IsOwner)
            {
                Reconciliation(default, false);
                CheckInput(out MoveData md);
                Move(md, false);
            }
            if (base.IsServer)
            {
                Move(default, true);
            }
        }


        private void TimeManager_OnPostTick()
        {
            if (base.IsServer)
            {
                var testObjectPos = (_testObject != null) ? _testObject.Position : Vector3.zero;
                ReconcileData rd = new ReconcileData(transform.position, transform.rotation, _rigidbody.velocity, _rigidbody.angularVelocity, testObjectPos);
                Reconciliation(rd, true);
            }
        }

        private void CheckInput(out MoveData md)
        {
            md = default;

            float horizontal = Input.GetAxisRaw("Horizontal");
            float vertical = Input.GetAxisRaw("Vertical");

            if (horizontal == 0f && vertical == 0f && !_jump)
                return;

            md = new MoveData(_jump, horizontal, vertical);
            _jump = false;
        }

        [Replicate]
        private void Move(MoveData md, bool asServer, bool replaying = false)
        {
            Vector3 forces = new Vector3(md.Horizontal, Physics.gravity.y, md.Vertical) * _moveRate;
            _rigidbody.AddForce(forces);

            if (md.Jump)
                _rigidbody.AddForce(new Vector3(0f, _jumpForce, 0f), ForceMode.Impulse);
        }

        [Reconcile]
        private void Reconciliation(ReconcileData rd, bool asServer)
        {
            transform.position = rd.Position;
            transform.rotation = rd.Rotation;
            _rigidbody.velocity = rd.Velocity;
            _rigidbody.angularVelocity = rd.AngularVelocity;

            // This will create invalid IL Code
            if (_testObject != null) {
                _testObject.Position = rd.ObjectPos;
            }
        }
    }
}```

**Comment by Lutz89** [comment]
**Winterbolt**
It definitely is because I'm getting the same de-compilation error as you
The errors all go away in 2022.1 though
So, this definitely has to do with Unity 2020 and 2021
The errors (in 2020 and 2021) happen with normal classes, abstract classes and interfaces

**Comment by Lutz89** [comment]
**Winterbolt**
It definitely is because I'm getting the same de-compilation error as you
The errors all go away in 2022.1 though
So, this definitely has to do with Unity 2020 and 2021
The errors (in 2020 and 2021) happen with normal classes, abstract classes and interfaces

**Comment by Abdelfattah-Radwan** [comment]
I can confirm this. I tested in 2020 & 2021 with classes, abstract classes, and interfaces, and all give the "Invalid branch target" when decompiling using ILSpy or an "Unable to decompile method" error when using dotPeek.

**Comment by ns2808** [comment]
I can confirm this as well.

**Comment by FirstGearGames** [comment]
I see the IL errors. Looking into this.

**Comment by FirstGearGames** [comment]
This screams Unity bug to me, but I'm investigating further.

The IL code is broken using your example, but if you put literally any logic after it works fine.

**Comment by FirstGearGames** [comment]
I found the issue. It seems the code may becoming optimized by the compiler which is presenting an unexpected pattern in my codegen, and it handles the changes incorrectly. Why it's not a problem in 2021 I'm not sure, but this is resolved in 1.5.6.
